lib/ansible/modules/crypto/openssl_csr.py

```diff
@@ -338,6 +353,9 @@ def generate(self, module):
                 usages = ', '.join(self.basicConstraints)
                 extensions.append(crypto.X509Extension(b"basicConstraints", self.basicConstraints_critical, usages.encode('ascii')))
 
+            if self.ocspMustStaple:
+                extensions.append(crypto.X509Extension(b"1.3.6.1.5.5.7.1.24", self.ocspMustStaple_critical, b"DER:30:03:02:01:05"))
```

Comment:
Where does this magic constant (the `DER:`) come from? I'm not seeing it in the RFC.

Reply:
ah it appears to be the "DER encoding of the Features structure", but I'm still not entirely sure what's encoded here. All these magic numbers are highly annoying :stuck_out_tongue_winking_eye: .

Reply:
You can find it both in https://scotthelme.co.uk/ocsp-must-staple/ and https://github.com/certbot/certbot/blob/d2df6880e5601af2813a7455abb265686a4c8b83/certbot/crypto_util.py#L123.

Reply:
:+1:

Reply:
Could we add some kind of openssl version detection to define variables. I am not a huge fan of magic number laying around. If we could have a variable that would meet both scenarios (pre-1.1 and post) this seems to me that it would be ideal and more comprehensive.

```python

import ssl

if ssl.OPENSSL_VERSION > 1.1.1:
  MUST_STAPLE_NAME = "tlsfeature"
  MUST_STAPLE_VALUE = "status_request"
else:
  MUST_STAPLE_NAME = "1.3.6.1.5.5.7.1.24"
  MUST_STAPLE_VALUE = "DER:30:03:02:01:05"

[...]

extensions.append(crypto.X509Extension(MUST_STAPLE_NAME, self.ocspMustStaple_critical, MUST_STAPLE_VALUE))
 ```

What do you guys think ?

Reply:
It may be helpful to clean up this line, but I don't think it's worth much unless it also allows to clean up `_check_ocspMustStaple`. I'm not sure what OpenSSL > 1.1 will report for a certificate generated with the "magic" number `1.3.6.1.5.5.7.1.24`. If it _does_ print it as `tlsfeature`, introducing constants such as `MUST_STAPLE_NAME` should also greatly simplify those lines of code.

Reply:
It does print `tlsfeature`, as the resulting certificate doesn't depend whether it was created with the magic number or `tlsfeature`. Both can be used to create a certificate with OpenSSL >= 1.1.

Reply:
@Spredzy: the version with your change still works, so it should be fine ;) I used `OpenSSL.SSL.OPENSSL_VERSION_NUMBER` as I wasn't sure whether `ssl` will always use the same OpenSSL library than `pyOpenSSL` (also, `OpenSSL` already has been imported).
@thomwiggers: it does clean up `_check_ocspMustStaple` a bit, so I think it should be ok.

In any case, if you prefer the old version, we can still get rid of the second commit :)